rust/src/crypto/encrypt.rs

```diff
@@ -0,0 +1,116 @@
+use super::ByteObject;
+
+use derive_more::{AsMut, AsRef, From};
+use sodiumoxide::crypto::{box_, sealedbox};
+
+/// Generate a new random key pair
+pub fn generate_encrypt_key_pair() -> (PublicEncryptKey, SecretEncryptKey) {
+    let (pk, sk) = box_::gen_keypair();
+    (PublicEncryptKey(pk), SecretEncryptKey(sk))
+}
+
+/// Public key for asymmetric authenticated encryption
+#[derive(
+    AsRef,
+    AsMut,
+    From,
+    Serialize,
+    Deserialize,
+    Hash,
+    Eq,
+    Ord,
+    PartialEq,
+    Copy,
+    Clone,
+    PartialOrd,
+    Debug,
+)]
+pub struct PublicEncryptKey(box_::PublicKey);
+
+impl ByteObject for PublicEncryptKey {
+    fn zeroed() -> Self {
+        Self(box_::PublicKey([0_u8; box_::PUBLICKEYBYTES]))
+    }
+
+    fn as_slice(&self) -> &[u8] {
+        self.0.as_ref()
+    }
+
+    fn from_slice(bytes: &[u8]) -> Option<Self> {
+        box_::PublicKey::from_slice(bytes).map(Self)
+    }
+}
+
+pub const SEALBYTES: usize = sealedbox::SEALBYTES;
+
+impl PublicEncryptKey {
+    /// Encrypt a message `m` with this public key. The resulting
+    /// ciphertext length is [`SEALBYTES`] + `m.len()`.
+    ///
+    /// The function creates a new key pair for each message, and
+    /// attaches the public key to the ciphertext. The secret key is
+    /// overwritten and is not accessible after this function returns.
+    pub fn encrypt(&self, m: &[u8]) -> Vec<u8> {
+        sealedbox::seal(m, self.as_ref())
+    }
+}
+
+impl SecretEncryptKey {
+    /// Decrypt the ciphertext `c` using this secret key and the
+    /// associated public key, and return the decrypted message.
+    ///
+    /// If decryption fails `Err(())` is returned.
+    pub fn decrypt(&self, c: &[u8], pk: &PublicEncryptKey) -> Result<Vec<u8>, ()> {
+        sealedbox::open(c, pk.as_ref(), self.as_ref())
+    }
+
+    /// Compute the corresponding public key for this secret key
+    pub fn public_key(&self) -> PublicEncryptKey {
+        PublicEncryptKey(self.0.public_key())
+    }
+}
+
+/// Secret key for asymmetric authenticated encryption
+#[derive(AsRef, AsMut, From, Serialize, Deserialize, Eq, PartialEq, Clone, Debug)]
+pub struct SecretEncryptKey(box_::SecretKey);
```

Comment:
it's probably better to have the definition of `SecretEncryptKey` before the impl in line 58